File: app/src/main/java/bjzhou/coolapk/app/net/ApiManager.java
package bjzhou.coolapk.app.net;

import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Build;
import android.text.Html;
import android.util.Log;

import com.jakewharton.retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import bjzhou.coolapk.app.model.Apk;
import bjzhou.coolapk.app.model.ApkField;
import bjzhou.coolapk.app.model.Comment;
import bjzhou.coolapk.app.model.UpgradeApkExtend;
import bjzhou.coolapk.app.util.Constant;
import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.ObservableSource;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;
import okhttp3.HttpUrl;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Created by bjzhou on 14-7-29.
 */
public class ApiManager {

    private static final String TAG = "ApiManager";

    private static ApiManager instance;
    private final CoolapkService mService;

    private ApiManager() {
        OkHttpClient client = new OkHttpClient.Builder().addInterceptor(new Interceptor() {
            @Override
            public Response intercept(Chain chain) throws IOException {
                Request original = chain.request();
                HttpUrl url = original.url().newBuilder()
                        .addQueryParameter("apikey", Constant.API_KEY)
                        .build();

                Request.Builder requestBuilder = original.newBuilder()
                        .header("Cookie", "coolapk_did=" + Constant.COOLAPK_DID)
                        .url(url);

                Request request = requestBuilder.build();
                return chain.proceed(request);
            }
        }).build();
        Retrofit retrofit = new Retrofit.Builder()
                .client(client)
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .addConverterFactory(GsonConverterFactory.create())
                .baseUrl(Constant.COOLAPK_PREURL)
                .build();

        mService = retrofit.create(CoolapkService.class);
    }

    public static ApiManager getInstance() {
        if (instance == null) {
            instance = new ApiManager();
        }
        return instance;
    }

    public static boolean isWifiConnected(Context context) {
        ConnectivityManager connManager = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo wifi = connManager.getNetworkInfo(ConnectivityManager.TYPE_WIFI);
        return wifi.isConnected();
    }

    public static boolean checkNetworkState(Context context) {
        ConnectivityManager cm = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
        return cm.getActiveNetworkInfo() != null && cm.getActiveNetworkInfo().isConnectedOrConnecting();
    }

    public Observable<List<Apk>> obtainHomepageApkList(int page) {
        return mService.obtainHomepageApkList(page)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Observable<ApkField> obtainApkField(int id) {
        return mService.obtainApkField(id)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Observable<List<Apk>> obtainSearchApkList(String query, int page) {
        return mService.obtainSearchApkList(query, page)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Observable<List<Comment>> obtainCommentList(int id, int page) {
        return mService.obtainCommentList(id, page)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Observable<List<UpgradeApkExtend>> obtainUpgradeVersions(Context context) {
        final PackageManager pm = context.getPackageManager();
        return Observable
                .create(new ObservableOnSubscribe<String>() {
                    @Override
                    public void subscribe(ObservableEmitter<String> e) throws Exception {
                        List<ApplicationInfo> all = pm.getInstalledApplications(0);
                        StringBuilder sb = new StringBuilder();
                        sb.append("sdk=").append(Build.VERSION.SDK_INT);
                        sb.append("&pkgs=");
                        for (ApplicationInfo info : all) {
                            if ((info.flags & ApplicationInfo.FLAG_SYSTEM) == 0) {
                                sb.append(info.packageName).append(",");
                            }
                        }
                        Log.d(TAG, "subscribe: " + sb);
                        e.onNext(sb.toString());
                        e.onComplete();
                    }
                })
                .subscribeOn(Schedulers.io())
                .concatMap(new Function<String, ObservableSource<List<Apk>>>() {
                    @Override
                    public ObservableSource<List<Apk>> apply(String s) throws Exception {
                        Log.d(TAG, "apply: obtainUpgradeVersions");
                        return mService.obtainUpgradeVersions(s);
                    }
                })
                .map(new Function<List<Apk>, List<UpgradeApkExtend>>() {
                    @Override
                    public List<UpgradeApkExtend> apply(List<Apk> apks) throws Exception {
                        List<UpgradeApkExtend> updateList = new ArrayList<UpgradeApkExtend>();
                        Log.d(TAG, "apply: onResponse");
                        for (Apk apk : apks) {
                            try {
                                PackageInfo pi = pm.getPackageInfo(apk.getApkname(), 0);
                                if (apk.getApkversioncode() > pi.versionCode) {
                                    UpgradeApkExtend ext = new UpgradeApkExtend();
                                    ext.setTitle(pi.applicationInfo.loadLabel(pm).toString());
                                    ext.setLogo(pi.applicationInfo.loadIcon(pm));
                                    String info = "<font color=\"#ff35a1d4\">" + pi.versionName + "</font>"
                                            + ">>"
                                            + "<font color=\"red\">" + apk.getApkversionname() + "</font>"
                                            + "<font color=\"black\">(" + apk.getApksize() + ")</font>";
                                    ext.setInfo(Html.fromHtml(info));
                                    ext.setApk(apk);
                                    updateList.add(ext);
                                }
                            } catch (PackageManager.NameNotFoundException ignored) {
                            }
                        }
                        return updateList;
                    }
                })
                .observeOn(AndroidSchedulers.mainThread());
    }
}

File: app/src/main/kotlin/bjzhou/coolapk/app/net/ApiManager.kt
package bjzhou.coolapk.app.net

import android.content.Context
import android.content.pm.ApplicationInfo
import android.net.ConnectivityManager
import android.os.Build
import android.text.Html
import android.util.Log
import bjzhou.coolapk.app.model.UpgradeApkExtend
import bjzhou.coolapk.app.util.Constant
import bjzhou.coolapk.app.retrofit.KCallAdapterFactory
import bjzhou.coolapk.app.retrofit.V6GsonConverterFactory
import bjzhou.coolapk.app.util.Utils
import com.coolapk.market.util.AuthUtils
import kotlinx.coroutines.experimental.Deferred
import kotlinx.coroutines.experimental.async
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.converter.scalars.ScalarsConverterFactory
import java.util.concurrent.TimeUnit

/**
 * Created by bjzhou on 14-7-29.
 */
class ApiManager private constructor() {
    lateinit var mService: CoolapkService
    lateinit var mServiceV6: CoolMarketServiceV6

    init {
        initService()
        initServiceV6()
    }

    private fun initServiceV6() {
        val client = OkHttpClient.Builder()
                .connectTimeout(5, TimeUnit.SECONDS)
                .readTimeout(15, TimeUnit.SECONDS)
                .addInterceptor({ chain ->
                    val original = chain.request()

                    Log.d(TAG, "url: " + original.url())

                    val requestBuilder = original.newBuilder()
                            .header("User-Agent", Utils.userAgent)
                            .header("X-Requested-With", "XMLHttpRequest")
                            .header("X-Sdk-Int", Build.VERSION.SDK_INT.toString())
                            .header("X-Sdk-Locale", Utils.localeString)
                            .header("X-App-Id", "coolmarket")
                            .header("X-App-Token", AuthUtils.getAS(Utils.uuid))
                            .header("X-App-Version", "7.3")
                            .header("X-App-Code", "1701135")
                            .header("X-Api-Version", "7")

                    val request = requestBuilder.build()
                    chain.proceed(request)
                }).build()
        val retrofit = Retrofit.Builder()
                .client(client)
                .addCallAdapterFactory(KCallAdapterFactory())
                .addConverterFactory(V6GsonConverterFactory.create())
                .baseUrl(Constant.COOL_MARKET_PREURL_V6)
                .build()
        mServiceV6 = retrofit.create(CoolMarketServiceV6::class.java)
    }

    private fun initService() {
        val client = OkHttpClient.Builder()
                .connectTimeout(5, TimeUnit.SECONDS)
                .readTimeout(15, TimeUnit.SECONDS)
                .addInterceptor { chain ->
                    val original = chain.request()
                    val url = original.url().newBuilder()
                            .addQueryParameter("apikey", Constant.API_KEY)
                            .build()

                    val requestBuilder = original.newBuilder()
                            .header("Cookie", "coolapk_did=" + Constant.COOLAPK_DID)
                            .url(url)

                    val request = requestBuilder.build()
                    chain.proceed(request)
                }.build()
        val retrofit = Retrofit.Builder()
                .client(client)
                .addCallAdapterFactory(KCallAdapterFactory())
                .addConverterFactory(GsonConverterFactory.create())
                .baseUrl(Constant.COOLAPK_PREURL)
                .build()

        mService = retrofit.create(CoolapkService::class.java)
    }

    fun obtainUpgradeVersions(context: Context): Deferred<List<UpgradeApkExtend>> {
        return async {
            val pm = context.packageManager
            val all = pm.getInstalledApplications(0)
            val sb = StringBuilder()
            sb.append("sdk=").append(Build.VERSION.SDK_INT)
            sb.append("&pkgs=")
            all
                    .filter { it.flags and ApplicationInfo.FLAG_SYSTEM == 0 }
                    .forEach { sb.append(it.packageName).append(",") }
            Log.d(TAG, "subscribe: $sb")
            val apks = mService.obtainUpgradeVersions(sb.toString()).execute() ?: emptyList()
            Log.d(TAG, "apply: onResponse")
            apks.filter {
                val pi = pm.getPackageInfo(it.apkname, 0)
                it.apkversioncode > pi.versionCode
            }.map {
                val pi = pm.getPackageInfo(it.apkname, 0)
                val ext = UpgradeApkExtend()
                ext.title = pi.applicationInfo.loadLabel(pm).toString()
                ext.logo = pi.applicationInfo.loadIcon(pm)
                val info = "<font color=\"#ff35a1d4\">" +
                        pi.versionName +
                        "</font>" + ">>" +
                        "<font color=\"red\">" +
                        it.apkversionname + "</font>" +
                        "<font color=\"black\">(" + it.apksize + ")</font>"
                ext.info = Html.fromHtml(info)
                ext.apk = it
                ext
            }
        }
    }

    companion object {

        private val TAG = ApiManager::class.java.simpleName

        var instance: ApiManager = ApiManager()
            private set

        fun isWifiConnected(context: Context): Boolean {
            val connManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
            val wifi = connManager.getNetworkInfo(ConnectivityManager.TYPE_WIFI)
            return wifi.isConnected
        }

        fun checkNetworkState(context: Context): Boolean {
            val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
            return cm.activeNetworkInfo != null && cm.activeNetworkInfo.isConnectedOrConnecting
        }
    }
}
